Build system changes for C <=> Rust interop

XrXr opened the pull request:
For reading: the doc comments. Feedback welcome so we can improve the
design together and be more aligned.

For playing with: the "works on my machine" demo.
 1. make sure you rerun the configure script with --enable-yjit=dev.
    I'd do a `make distclean` first and _not_ use `./configure -C` caching.
 2. `make -j miniruby`
 3. `./miniruby --yjit` to see the output of the demo.

There is of course stuff in here that we eventually want to merge but
I haven't put in the effort to cherry pick stuff out and refine
important bits.

Sorry about the long essay and invitations to read stuff I haven't fully
read myself.

noahgibbs commented:
This doesn't seem to be enough to actually link, though. Or maybe it is, but it's not in the Makefile?

noahgibbs commented:
I did the distclean and full reconfigure and clean. Here's what I'm currently seeing:

~~~
linking miniruby
Undefined symbols for architecture x86_64:
  "_Init_builtin_yjit", referenced from:
      _rb_call_builtin_inits in inits.o
  "_rb_yjit_before_ractor_spawn", referenced from:
      _ractor_create in ractor.o
  "_rb_yjit_bop_redefined", referenced from:
      _rb_vm_check_redefinition_opt_method in vm.o
  "_rb_yjit_call_threshold", referenced from:
      _mjit_exec in vm.o
  "_rb_yjit_cme_invalidate", referenced from:
      _clear_method_cache_by_id_in_class in vm.o
  "_rb_yjit_compile_iseq", referenced from:
      _mjit_exec in vm.o
  "_rb_yjit_constant_ic_update", referenced from:
      _vm_exec_core in vm.o
  "_rb_yjit_constant_state_changed", referenced from:
      _rb_clear_constant_cache in vm.o
  "_rb_yjit_enabled_p", referenced from:
      _Init_ruby_description in version.o
      _ruby_show_version in version.o
      _mjit_exec in vm.o
  "_rb_yjit_init", referenced from:
      _ruby_process_options in ruby.o
  "_rb_yjit_invalidate_all_method_lookup_assumptions", referenced from:
      _rb_clear_method_cache_all in vm.o
  "_rb_yjit_iseq_free", referenced from:
      _rb_iseq_free in iseq.o
  "_rb_yjit_iseq_mark", referenced from:
      _rb_iseq_mark in iseq.o
  "_rb_yjit_iseq_update_references", referenced from:
      _rb_iseq_update_references in iseq.o
  "_rb_yjit_method_lookup_change", referenced from:
      _clear_method_cache_by_id_in_class in vm.o
  "_rb_yjit_tracing_invalidate_all", referenced from:
      _rb_hook_list_free in vm_trace.o
      _rb_thread_add_event_hook in vm_trace.o
      _rb_add_event_hook2 in vm_trace.o
      _rb_thread_add_event_hook2 in vm_trace.o
      _tracepoint_enable_m in vm_trace.o
      _clean_hooks_check in vm_trace.o
      _thread_add_trace_func in vm_trace.o
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [miniruby] Error 1
~~~

noahgibbs commented:
Ah, never mind. Left off an arg where I didn't think I did. It works, thank you!

maximecb commented on yjit/src/yjit.rs:
```diff
@@ -1,11 +1,135 @@
+//  - Using rust-bindgen this way seems unusual. We might be depending on parts
+//    that the project is not committed to maintaining
+//  - This setup assumes rust-bindgen gives deterministic output, which can't be taken
+//    for granted
```
Non-deterministic output from bindgen would seem like a bug.

XrXr replied:
My worry here is that the set of outputs would all be semantically equivalent, but not bit by bit identical. For example if it used a hash map internally it might output things in different orders. The recommended way to use the library is apparently running it inside the build script: <https://rust-lang.github.io/rust-bindgen/library-usage.html>. Complete determinism for that use case doesn't seem important.

noahgibbs replied:
Definitely possible. Presumably we'll find out!

maximecb replied:
We probably should use bindgen in the intended way and run it inside the build script IMHO. More robust and would presumably enable us to export C struct layouts as well.

XrXr replied:
> We probably should use bindgen in the intended way and run it inside the build script IMHO. More robust and would presumably enable us to export C struct layouts as well.

I disagree. Let's break this down. It's not necessarily more robust because it's an extra `crates.io` build dependency, so right off the bat that requires internet to build YJIT and makes packaging for distros more difficult. Additionally, rust-bindgen depends on the system Clang through `clang-sys`, so that's yet another variable and point of failure in the build process. We can't lock the Clang version that bindgen with Cargo. Comparatively, we can review the output binding file easily, and it'll be version controlled. When we want to add new bindings it's more cumbersome than having it done automagically, but as a silver lining, requiring an explicit step makes us think twice before depending on stuff from the rest of the codebase that have interfaces which are not so well defined. 

Running inside build.rs would only allow us to robustly export C struct layouts if we are also willing to require that the rest of CRuby is built using the same Clang. That's an onerous requirement. Version controlling the output allows us to have access to struct layouts as well, though we'll need to limit ourselves to structs that are not build configuration dependent.

Using bindgen the way I'm proposing is common enough that we have good company. Fuchsia seems to recommend this usage https://fuchsia.dev/fuchsia-src/development/languages/rust/bindgen.

maximecb replied:
I thought that people packaging for distros were shipping compiled binaries (as is the case on Ubuntu). Is the problem that some distros build packages from source on the client side? I can see that being a much bigger problem if so.

> Running inside build.rs would only allow us to robustly export C struct layouts if we are also willing to require that the rest of CRuby is built using the same Clang.

Is that really true? Does the C ABI not specify how structs should be laid out? Seems like the kind of problem that should have been solved decades ago.

> Using bindgen the way I'm proposing is common enough that we have good company. Fuchsia seems to recommend this usage https://fuchsia.dev/fuchsia-src/development/languages/rust/bindgen.

It's reassuring that we're not the only ones doing it this way then 👍

XrXr replied:
>  Is the problem that some distros build packages from source on the client side?

Yes, that's part of the problem. Adding complexity to the build process makes it harder for packagers to run their CI and for source based distros means impacting end users. For development versions, there is often no choice but to build from source.

> Is that really true? Does the C ABI not specify how structs should be laid out? Seems like the kind of problem that should have been solved decades ago.

I don't think it's completely solved by the ABI. One possible failure mode is that through feature detection headers expose semantically different API based on compiler feature detection. Consider this  https://github.com/ruby/ruby/blob/master/include/ruby/internal/stdalign.h#L53-L73 on an old GCC but a recent Clang for example.

If this ends up causing bugs it can be hard for us to reproduce because build.rs has output we don't control tied to the specific system configuration the person building from source has.

Sidenote regarding "this is kind of a weird way to use bindgen", I just realized that a lot of the binding crates on crates.io do what we do by publishing bindgen output to crates.io. Two examples are <https://crates.io/crates/curl-sys> and <https://github.com/capstone-rust/capstone-rs/tree/master/capstone-sys>. I think there are a lot of these `*-sys` binding crates!

noahgibbs replied:
While there are definitely ways to choose different struct layouts (different per-compiler features, #ifdef, etc), overall C struct layout is quite stable from compiler to compiler. C++ never really got this figured out, but C struct layout is used by a variety of languages for interop, and is overall quite stable and well-established. It was already surprisingly solid when I first learned about this in 1995-ish. The main subtlety is with field alignment, and it's not a huge subtlety.

Certainly running bindgen early and just checking in the output sounds reasonable to me. Makes sense that that's a common approach. It won't 100% solve any possible problems with determinism, but it'll make them a lot less annoying for us.

maximecb commented on yjit/src/yjit.rs:
```diff
@@ -1,11 +1,135 @@
+//
+// TODO(alan): is the model different enough on Windows that this setup is unworkable?
+//             Seems prudent to at least learn more about Windows binary tooling before
+//             committing to a design.
```
Glad to see you're thinking about things like this 🙏

maximecb commented on yjit.h:
```diff
@@ -62,5 +78,29 @@ void rb_yjit_iseq_free(const struct rb_iseq_constant_body *body);
+inline void rb_yjit_iseq_free(const struct rb_iseq_constant_body *body) {}
+inline void rb_yjit_before_ractor_spawn(void) {}
+inline void rb_yjit_constant_ic_update(const rb_iseq_t *const iseq, IC ic) {}
+inline void rb_yjit_tracing_invalidate_all(void) {}
```
These are all the functions we'll export from Rust?

XrXr replied:
Yeah, most of these are going to be exported from Rust.

XrXr committed:
Link C YJIT and Rust YJIT

 - C YJIT is only built in `./configure --enable-yjit` builds unlike
   upstream, where stub implementations for yjit.h functions come
   from the always built yjit.o unit.
 - `./miniruby --yjit` runs some Rust code. Please compare output with
   source
 - Functions that Rust YJIT wants to call go into `crate::cruby` for
   now. We may want a dedicated module for YJIT functions written in C.

XrXr commented:
I'm going to merge this for now as discussed since the build system changes make CI more green and unblocks some work that require calling C code. The binding generation tool will be done separately and we can write signatures by hand in the meantime.

maximecb commented:
Good job Alan. Thanks for merging this :)

I think the most pressing thing for us next would be auto-exporting all the constants we need. In particular, `SIZEOF_` and `OFFSET_` for the structs we need access to. Like you said we can write function signatures by hand, but manually exporting field offsets seems super error-prone and brittle.
